server/src/main/java/org/elasticsearch/search/aggregations/pipeline/InternalPercentileRanksBucket.java

```diff
@@ -0,0 +1,202 @@
+        if ((PercentileRanks.length == percents.length) == false) {
+            throw new IllegalArgumentException("The number of provided percents and PercentileRanks didn't match. percents: "
+                    + Arrays.toString(percents) + ", Percentile Ranks: " + Arrays.toString(PercentileRanks));
+        }
+        this.format = formatter;
+        this.percentile_ranks = percentile_ranks;
+        this.percents = percents;
+        this.keyed = keyed;
+        computeLookup();
+    }
+
+    private void computeLookup() {
+        for (int i = 0; i < percents.length; i++) {
+            percentileLookups.put(percents[i], percentile_ranks[i]);
+        }
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public InternalPercentileRanksBucket(StreamInput in) throws IOException {
+        super(in);
+        format = in.readNamedWriteable(DocValueFormat.class);
+        percentile_ranks = in.readDoubleArray();
+        percents = in.readDoubleArray();
+        keyed = in.readBoolean();
+
+        computeLookup();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(format);
+        out.writeDoubleArray(percentile_ranks);
+        out.writeDoubleArray(percents);
+        out.writeBoolean(keyed);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return PercentileRanksBucketPipelineAggregationBuilder.NAME;
+    }
+
+    @Override
+    public double percentile_rank(double percent) throws IllegalArgumentException {
+        Double percentile_ranks = percentileLookups.get(percent);
+        if (percentile_ranks == null) {
+            throw new IllegalArgumentException("Percent requested [" + String.valueOf(percent) + "] was not" +
+                    " one of the computed percentile ranks.  Available keys are: " + Arrays.toString(percents));
+        }
+        return percentile_ranks;
+    }
+
+    @Override
+    public String percentileRankAsString(double percent) {
+        return format.format(percentile_rank(percent)).toString();
+    }
+
+    DocValueFormat formatter() {
+        return format;
+    }
+
+    @Override
+    public Iterator<Percentile> iterator() {
+        return new Iter(percents, percentile_ranks);
+    }
+
+    @Override
+    public double value(String name) {
+        return percentile_rank(Double.parseDouble(name));
+    }
+
+    @Override
+    public InternalMax doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        throw new UnsupportedOperationException("Not supported");
+    }
+
+    @Override
+    public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        if (keyed) {
+            builder.startObject("values");
+            for (double percent : percents) {
+                double value = percentile_rank(percent);
+                boolean hasValue = !(Double.isInfinite(value) || Double.isNaN(value));
```

Comment:
Style nit: we've stopped using `!(foo)` negations since they are easy to misread, and instead prefer the longer `foo != true` or `foo == false`.

(I know this was copied from the percentiles_bucket agg, it has the older style we are moving away from :) )

Reply:
Ahh. Sure. I agree. `foo != true` seems more readable code.